### PR TITLE
Introduce interpreter implementation

### PR DESCRIPTION
**Pull request by dmlloyd**
Introduce interpreter implementation, and add a command-line switch to enable build time class initialization.

### CODE REVIEW

File: interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java
Comment (DanHeidinga) [inline interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: Does this globally intern the string?  All string constants need to be interned but we can add that as a follow on task as well

File: interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java
Comment (DanHeidinga) [inline interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1666]: We probably want this to be wrapped in an api like `getClassLoaderIncludingBoot()` (with a better name) so we don't have to sprinkle the null check throughout the VM code.

File: interpreter/src/main/java/org/qbicc/interpreter/impl/VmClassLoaderImpl.java
Comment (DanHeidinga) [inline interpreter/src/main/java/org/qbicc/interpreter/impl/VmClassLoaderImpl.java:28]: This should move the Vm class so it can be shared by all VmClassLoaderImpl

File: interpreter/src/main/java/org/qbicc/interpreter/impl/VmClassLoaderImpl.java
Comment (DanHeidinga) [inline interpreter/src/main/java/org/qbicc/interpreter/impl/VmClassLoaderImpl.java:101]: I think we also need to record the successfully loaded classes in this classloader's `classes` hashtable to ensure we don't attempt to load the class a second time if we've successfully loaded it once

**Review by DanHeidinga — COMMENTED** [review]
Some initial comments.  Overall, the approach seems reasonable and the apis look usable.

File: interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java
Comment (dgrove-oss) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: The BuildTimeHeap code will intern StringLiterals when it emits them into the native image.  Maybe that is good enough (if we define compile time `equals` between StringLiteral values appropriately)?

Comment (DanHeidinga) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: User code can trip up on this though.... with the interpreter not recognizing two strings as `==` that should be, or with the final native image treating two strings as `==` when it shouldn't.  It's annoying, but I think we need to follow the programmer's intent on this.

The BuildTimeHeap can intern common `char[]'s from strings with the same content but different identity, but it probably shouldn't remove the identity differences.

Comment (dgrove-oss) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: In BuildTimeHeap, I'm interring StringLiteral but not interning instances of java.lang.String that come through as ObjectLiterals.  My assumption was that a StringLiteral would only be coming from a string literal in a class file, and therefore should be interned by the system (while ObjectLiteral would be interned by the program).

Comment (DanHeidinga) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: You're right about BuildTimeHeap being OK to intern `.equals` `StringLiterals`.  The only code I can find that creates `StringLiterals` is in ClassFileImpl and only for constantpool entries.

The interpreter may still treat two intern'd Strings as different if they originated in different classloaders so this still needs to be addressed or `==` can change results between interpreter and compiled image

File: compiler/src/main/java/org/qbicc/interpreter/VmObject.java
Comment (dmlloyd) [reply compiler/src/main/java/org/qbicc/interpreter/VmObject.java:13]: It's a per-object `Memory`, so it is GC'd independently from other objects using the host GC.

File: interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java
Comment (dmlloyd) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: Yeah only string literals are intentionally interned, and certain other specific cases like class names and so forth.  I do agree that string literals should be interned on a per-class-loader basis.

File: interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java
Comment (dmlloyd) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1666]: Yeah I agree, this needs to be improved.

File: interpreter/src/main/java/org/qbicc/interpreter/impl/VmClassLoaderImpl.java
Comment (dmlloyd) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/VmClassLoaderImpl.java:101]: Yeah this is a mess right now, and in fact this problem is the reason this is still in draft.  I'm trying out a couple different approaches and we'll see what fits best.

File: interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java
Comment (DanHeidinga) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: > The interpreter may still treat two intern'd Strings as different if they originated in different classloaders so this still needs to be addressed or `==` can change results between interpreter and compiled image

Sorry if I wasn't clear - this interpreter behaviour is a bug.  per-class-loader interning of strings is contrary to the spec

Comment (dmlloyd) [reply interpreter/src/main/java/org/qbicc/interpreter/impl/Frame.java:1050]: > > The interpreter may still treat two intern'd Strings as different if they originated in different classloaders so this still needs to be addressed or `==` can change results between interpreter and compiled image
> 
> Sorry if I wasn't clear - this interpreter behaviour is a bug. per-class-loader interning of strings is contrary to the spec

Ah okay, well good to know.  I'll get that fixed...

**Comment by dmlloyd** [comment]
Round 2.  Distinct subclasses for each kind of `Class`, more distinct subclasses for different kinds of `Object` with better capabilities.  Fixed class loading, more or less.  Lots of cleanup.

**Comment by dmlloyd** [comment]
Some exciting progress: the interpreter is now running class initialization. There is still much to do but this is major progress.

**Comment by dmlloyd** [comment]
This is now ready for full review. All classes without exception are build-time initialized for now. Enabling the `--init-build-time` switch does result in compilation failure - the majority of classes will fail to initialize due to missing native methods, intrinsics that are not designed for compile time, etc., however these problems need to be tackled incrementally.

File: plugins/layout/src/main/java/org/qbicc/plugin/layout/Layout.java
Comment (dgrove-oss) [inline plugins/layout/src/main/java/org/qbicc/plugin/layout/Layout.java:49]: Is the separate layout for the interpreter just used for static fields?  Or is it used to account for differences in the instance fields between qbicc and hostVM classes (where such differences exist)?

File: compiler/src/main/java/org/qbicc/graph/ValueHandleVisitorLong.java
Comment (dgrove-oss) [inline compiler/src/main/java/org/qbicc/graph/ValueHandleVisitorLong.java:6]: I didn't see where this new visitor was used.  What am I missing?

**Review by dgrove-oss — APPROVED** [review]
Overall makes sense to me.

I believe this PR is building up a "heap" of VmObject instances in the Memory objects associated with the various initialized classes, but that this heap is not yet being written to the executable.  After this is merged, and in parallel with continuing to enhance the interpreter functionality, work can be done to BuildTimeHeap to replace the current set of reflective operations to read the host-VM objects with operations that instead use VmObject APIs to access the Memory for each build-time initialized object and write its contents into initialHeap.ll.  Is that correct, or did I miss something.

File: compiler/src/main/java/org/qbicc/graph/ValueHandleVisitorLong.java
Comment (dmlloyd) [reply compiler/src/main/java/org/qbicc/graph/ValueHandleVisitorLong.java:6]: It's used by `org.qbicc.interpreter.impl.Frame#getOffset` to get a numeric result for a handle (in this case, what the offset is for the member compared to the base of the `Memory` object) without having to construct a box for the value.

File: plugins/layout/src/main/java/org/qbicc/plugin/layout/Layout.java
Comment (dmlloyd) [reply plugins/layout/src/main/java/org/qbicc/plugin/layout/Layout.java:49]: The difference here is that it's possible for fields to be injected that the interpreter does not know about because the interpreter may need the layout info at an earlier point than would have otherwise been determined. It's possible we could reconcile this at some point though - I've ended up needing to force the injection of most of the interesting injected fields earlier anyway.

The static fields layout method is only used by the interpreter and the results are not cached so the actual Layout instance is not of particular importance for this use case.

File: compiler/src/main/java/org/qbicc/graph/ValueHandleVisitorLong.java
Comment (dgrove-oss) [reply compiler/src/main/java/org/qbicc/graph/ValueHandleVisitorLong.java:6]: Got it; thanks!  Missed that in the 2132 lines & GitHub had helpfully elided the details of the big diffs when I was doing a search.

File: plugins/layout/src/main/java/org/qbicc/plugin/layout/Layout.java
Comment (dgrove-oss) [reply plugins/layout/src/main/java/org/qbicc/plugin/layout/Layout.java:49]: thanks.  Makes sense.

**Comment by dgrove-oss** [comment]
> That is my intention for how it would work. Maybe we should also look into dropping `StringLiteral` in favor of `ObjectLiteral` as well.

Yeah.  It seems like a easily separable first step would be changing the `StringLiteral.value` field from `java.lang.String` to `VmString`.

**Comment by dmlloyd** [comment]
Do not merge yet; the integration tests are failing but not reporting the failure for some reason.

**Comment by dmlloyd** [comment]
I can't determine why the reachability graph is so much smaller. It may be due to a change in timing or something more nefarious but as the results have been consistent and correct AFAICT, I'm fine with this going in unless anyone has an objection.

**Comment by ashu-mehra** [comment]
@dmlloyd  why do we have the build time init flag disabled by default?

**Comment by dmlloyd** [comment]
> @dmlloyd why do we have the build time init flag disabled by default?

Because if it's enabled, compilation always fails with multiple exceptions in initializers. I will follow up with initializer fixes, and once "hello world" is working with build time init, we can switch the default.

**Pull request merged**
